packages/context-editor/src/plugins/attributePanel.ts

```diff
@@ -1,19 +1,195 @@
-import { Plugin, PluginKey } from "prosemirror-state";
+import type { PluginViewSpec, WidgetDecorationFactory } from "@prosemirror-adapter/core";
+import type { ReactPluginViewUserOptions } from "@prosemirror-adapter/react";
+import type { Transaction } from "prosemirror-state";
+import type { Decoration } from "prosemirror-view";
+
+import { EditorState, Plugin, PluginKey } from "prosemirror-state";
+import { DecorationSet } from "prosemirror-view";
 
 import type { PanelProps } from "../ContextEditor";
+import { AttributePanel } from "../components/AttributePanel";
+
+export const ATTRIBUTE_PANEL_ID = "context-editor-attribute-panel-container";
+
+export const attributePanelKey = new PluginKey<{
+	panelPosition: PanelProps;
+	setPanelPosition: React.Dispatch<React.SetStateAction<PanelProps>>;
+}>("panel");
+
+type AttributePanelState = {
+	panelPosition: PanelProps;
+	setPanelPosition: React.Dispatch<React.SetStateAction<PanelProps>>;
+	decorations?: DecorationSet;
+};
 
-export const attributePanelKey = new PluginKey("panel");
 export default (
 	panelPosition: PanelProps,
-	setPanelPosition: React.Dispatch<React.SetStateAction<PanelProps>>
+	setPanelPosition: React.Dispatch<React.SetStateAction<PanelProps>>,
+	getToggleWidget: WidgetDecorationFactory,
+	pluginViewFactory: (options: ReactPluginViewUserOptions) => PluginViewSpec
 ) => {
-	return new Plugin({
-		key: attributePanelKey,
-		state: {
-			init: () => {
-				return { panelPosition, setPanelPosition };
+	return [
+		new Plugin<AttributePanelState>({
+			key: attributePanelKey,
+			state: {
+				init: () => {
+					return { panelPosition, setPanelPosition };
+				},
+				apply: (tr, pluginState, prevEditorState, editorState) => {
+					return {
+						...(tr.getMeta(attributePanelKey) || pluginState),
+						decorations: getUpdatedDecorations(
+							tr,
+							pluginState,
+							prevEditorState,
+							editorState,
+							getToggleWidget
+						),
+					};
+				},
+			},
+			props: {
+				decorations(state) {
+					// return this.getState(state)?.decorations;
```

Comment:
Uncommenting this should use our state logic to only rerender decorations in parts of the doc that change instead of crawling the whole doc each time, [as recommended by marijn](https://discuss.prosemirror.net/t/any-tips-for-caching-improving-performance-of-decorations/8044/1). But even without that, memoization makes this fast enough to work for now.